### PR TITLE
LF: enforce non-empty maintainers in contract key lookup

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This fixes a bug in the Speedy interpreter.  With this change, the
interpreter crashes if it attempts to lookup using a contract key that
has an empty set of maintainers.

This is a breaking change approved by @bame-da .

CHANGELOG_BEGIN
- [LF] (Bug fix) enforce non-empty maintainers in contract key during
  lookupByKey.

CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/scenario-service/protos/scenario_service.proto [hunk 1]
```diff
@@ -167,8 +167,10 @@ message ScenarioError {
 
   message EmptyContractKeyMaintainers {
     Identifier template_id = 1;
-    Value arg = 2;
-    Value key = 3;
+    Value key = 2;
+    // Optional location of the update expression leading
+    // to the violation.
+    Location location = 3;
```
Comment (cocreature) [inline compiler/scenario-service/protos/scenario_service.proto:173]: What makes this one special? We don’t have locations for any of the other errors afaik.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala [hunk 2]
```diff
@@ -1371,6 +1371,31 @@ class EngineTest
...
+          ledgerTime = now,
+          submissionTime = now,
+          seeding = InitialSeeding.TransactionSeed(seed),
+          globalCids = Set.empty,
+        )
+        .consume(_ => None, lookupPackage, lookupKey)
+
+      result shouldBe 'left
+      val Left(err) = result
+      err.msg should include("Update failed due to a contract key with an empty sey of maintainers")
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:1397]:
```suggestion
      inside (result) {
        case Left(err) =>  err.msg should include("Update failed due to a contract key with an empty sey of maintainers")
      }
```

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala [hunk 3]
```diff
@@ -1634,7 +1659,7 @@ class EngineTest
       err.msg should include("precondition violation")
     }
 
-    "not be create if has an empty set of maintainer" in {
+    "crash if uses a key with an empty set of maintainers" in {
```
Comment (cocreature) [inline daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala:1662]: This is still a create no? Seems a bit confusing to rename this to basically the same name as the test above when one is about create and the other is about lookup.

File: compiler/scenario-service/protos/scenario_service.proto (hunk 1 above)
Comment (cocreature) [reply compiler/scenario-service/protos/scenario_service.proto:173]: And afaik we also always include the last location anyway so I’m not sure when this provides extra information.

**Review by cocreature — APPROVED** [review]
Nice, thank you!

File: compiler/scenario-service/protos/scenario_service.proto (hunk 1 above)
Comment (remyhaemmerle-da) [reply compiler/scenario-service/protos/scenario_service.proto]: dropped.

File: daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/engine/src/test/scala/com/digitalasset/daml/lf/engine/EngineTest.scala]: reverted.

File: compiler/damlc/tests/daml-test-files/AAEmptyContractKeyMaintainers.daml [hunk 4]
```diff
@@ -0,0 +1,24 @@
+-- @ERROR Attempt to operate on a contract key with an empty set of maintainers
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply compiler/damlc/tests/daml-test-files/AAEmptyContractKeyMaintainers.daml]: I applied your suggestion.

File: compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs [hunk 5]
```diff
@@ -230,15 +230,17 @@ prettyScenarioErrorError (Just err) =  do
...
-        "Attempt to create a contract key with an empty set of maintainers in:"
-          $$ nest 2
-          (   "create"
-          <-> prettyMay "<missing template id>" (prettyDefName world) scenarioError_EmptyContractKeyMaintainersTemplateId
-          $$ (   keyword_ "with"
-              $$ nest 2 (prettyMay "<missing argument>" (prettyValue' False 0 world) scenarioError_EmptyContractKeyMaintainersArg)
-             )
-          )
+      pure $ vcat
+        [ "Attempt to create or operate on a contract key with an empty set of maintainers"
```
Comment (hurryabit) [inline compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs:234]: Would it make sense to distinguish between creating a contract and looking up a contract in this error message?

**Review by hurryabit — COMMENTED** [review]
Is this actually the right behavior? Always returning `None` also seems reasonable. If there are no maintainers, there can't be a contract with that key and all the zero maintainers have authorized that lookup. cc @bame-da

**Comment by remyhaemmerle-da** [comment]
> Is this actually the right behavior? Always returning `None` also seems reasonable. If there are no maintainers, there can't be a contract with that key and all the zero maintainers have authorized that lookup. cc @bame-da

I quoted here an internal discussion between @andreaslochbihler-da and @bame-da.

@andreaslochbihler-da wrote: 
> What should happen if lookupByKey is called for a key with empty maintainers? #7597 forbids empty maintainers in create nodes, but we still allow empty maintainers in lookupByKey. A LookupByKey node with empty maintainers has no required authorizers and no informees, but it's nevertheless included in the transaction tree for the witnesses.
> Canton rejects such transactions because we've built on the assumption that every node in the transaction has at least one > informee. Can we avoid having to deal with the special case of a node without informees and just have DAML engine crash if LookupByKey is called with a key without maintainers, just like it is going to do for contract creation?

@bame-da answered: 
> Yes, that sounds like the consistent thing to do

**Review by andreaslochbihler-da — APPROVED** [review]
There is the question whether we want to support fetchByKey with empty maintainers though. For @bame-da 's use case of being able to upgrade an existing contract with empty maintainers, it'd be enough to support empty maintainers only in Fetch/Exercise nodes that do not come from `fetchByKey` / `exerciseByKey`. All `byKey` operations would then check for empty maintainers. IIUC this should be doable once we explicitly mark which nodes come from key operations. I've heard of plans for doing this.

**Review by hurryabit — APPROVED** [review]
Thanks for explaining why we want the behavior implemented in this change. I fully understand the rationale now and totally agree with it.

Thanks for implementing the change!

File: compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs (hunk 5 above)
Comment (remyhaemmerle-da) [reply compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs]: I created two exceptions.

**Pull request merged**